operator: improve AWS API pagination with automatic fallback

pillai-ashwin opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer's Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

<!-- Description of change -->

Deprecates the `--aws-pagination-enabled` flag and improves aws api pagination handling. Will remove `MaxResults=1000` parameter and let aws decide optimal page sizes for better performance.

## Background

The `--aws-pagination-enabled` flag (#39543) controls whether we set `MaxResults=1000` on aws api calls. When enabled (default), every call is limited to 1000 results - if you have 3K+ ENIs, that's 3+ separate api calls. When disabled, aws picks the page size, potentially using fewer calls.

The problem: most users don't understand when to use this flag. The default is conservative (pagination enabled), so everyone pays the performance cost even in small environments.

## Current implementation (needs revision)

The current pushed code implements a retry mechanism based on the suggestion from [#40443](https://github.com/cilium/cilium/issues/40443): try an unpaginated request first, and if it fails, retry with `MaxResults=1000`. The idea was to get the performance benefit of fewer api calls in the happy case while falling back to pagination on errors.

issues with this approach:
- unconditional retry on any error is anti-pattern (rate limiting, unnecessary retries)
- adds complexity without clear benefit

## Revised approach

Based on review feedback, will simplify to just remove the retry logic and omit `MaxResults` entirely.

Remove `MaxResults` parameter from 4 functions controlled by the flag (from commits [8b24557f36](https://github.com/cilium/cilium/commit/8b24557f36) and [479bdf55ae](https://github.com/cilium/cilium/commit/479bdf55ae)):
- `GetDetachedNetworkInterfaces`
- `describeNetworkInterfaces` 
- `describeNetworkInterfacesFromInstances`
- `describeSecurityGroups`

The sdk paginator still handles pagination and we're just not capping pages to 1000.

Deprecate the `--aws-pagination-enabled` flag (hidden, no-op) in v1.19, remove in v1.20.

## Safety

aws docs say `DescribeNetworkInterfaces` can fail with `OperationNotPermitted` if you request too many results without filters. We use `subnet-id` filters in all calls, which satisfies aws's requirement. Same for `DescribeSecurityGroups` - we pass vpc-id from instance metadata.

The original timeout issues ([479bdf55ae](https://github.com/cilium/cilium/commit/479bdf55ae)) happened because `DescribeSecurityGroups` was called with no filters and no `MaxResults`. Now we have filters, so result sets are already bounded.

## Expected improvement

Current: 3K+ ENIs = 3+ api calls with `MaxResults=1000`  
Expected: Same data in fewer calls if aws uses larger page sizes

Not sure what the default page size is when `MaxResults` isn't set but it should be at least more than 1000 to see perf improvements.

## Backwards compatibility

Flag is ignored but still accepted for v1.19, so existing deployments won't break. Users should remove it before v1.20.

## Testing

Existing tests pass

Fixes #40443

```release-note
operator: Improve AWS API pagination with adaptive fallback
```

tklauser commented on operator/option/config.go:
```diff
@@ -385,7 +386,8 @@ type OperatorConfig struct {
-	// AWSPaginationEnabled toggles pagination for AWS EC2 API requests
+	// Deprecated: AWSPaginationEnabled is deprecated and a no-op in v1.19.
+	// It will be removed in v1.20. Pagination now happens automatically.
 	AWSPaginationEnabled bool
```
I think this (and the corresponding assignment in `(*OperatorConfig).Populate` can already be removed since it is only set but never read after this PR.

tklauser reviewed (approved):
Thanks

qmonnet commented:
> I will not be able to look into this until I'm back from PTO. I don't know how to get someone else from AWS team to review it

Cc @cilium/aws

gandro commented on pkg/aws/ec2/ec2.go:
```diff
@@ -222,16 +221,36 @@ func (c *Client) GetDetachedNetworkInterfaces(ctx context.Context, tags ipamType
+	output, err := paginator.NextPage(ctx)
+	c.metricsAPI.ObserveAPICall(DescribeNetworkInterfaces, deriveStatus(err), sinceStart.Seconds())
+
+	if err != nil {
```
Unconditionally re-trying every request seems like an anti-pattern to me. Especially if the request fails because the API is rate-limiting us we will be making the problem worse. Can we detect that the error was due to unpaginated request? When would that even happen?

gandro commented:
> In the sdk docs vendor/github.com/aws/aws-sdk-go-v2/service/ec2/api_op_DescribeNetworkInterfaces.go - aws only throws OperationNotPermitted when you query too many resources without narrowing filters.

Are we really certain this is really the case? The way I read the upstream docs, `OperationNotPermitted` might also happen if you request too many resources without pagination, *even if filters are applied*

Is there a way to validate this assumption? I really don't want to accidentally undo what was done in https://github.com/cilium/cilium/commit/479bdf55ae

pillai-ashwin commented:
@gandro

> Allowing it to be configured sounds good to me. How would one disable pagination with that flag? Set it to zero?

Yeah, 0 would mean unset MaxResults and we can fallback to let AWS decide:
- `--aws-max-results-per-call=0` --> no MaxResults (AWS picks page size)  
- `--aws-max-results-per-call=1000`--> explicit page size (current default behavior)
- `--aws-max-results-per-call=5000`--> larger pages example

For backwards compat, keep `--aws-pagination-enabled` working:
- `--aws-pagination-enabled=true` maps to `--aws-max-results-per-call=1000`
- `--aws-pagination-enabled=false` maps to `--aws-max-results-per-call=0`

So users can keep using the old flag, but the new one gives more control. We can deprecate and remove the old flag in subsequent versions.

pillai-ashwin committed:
operator: add adaptive AWS API pagination

Defaults to unpaginated requests (fast) and automatically switches to
paginated (MaxResults=1000) when AWS returns OperationNotPermitted,
indicating too many results. The switch persists for the operator's
lifetime (until restart) and only retries the failed request once.

Replaces boolean --aws-pagination-enabled with int32
--aws-max-results-per-call flag for fine-grained control. Users can
set it to 0 (default, unpaginated with automatic fallback), 1000
(always paginated) or any other value for custom page sizes.

Fixes: #40443
Signed-off-by: Ashwin Pillai <pillaiashwin96@gmail.com>

pillai-ashwin commented:
/test

pillai-ashwin commented:
/test

gandro commented on operator/cmd/provider_aws_flags.go:
```diff
@@ -50,7 +50,13 @@ func (hook *awsFlagsHooks) RegisterProviderFlag(cmd *cobra.Command, vp *viper.Vi
 	option.BindEnv(vp, operatorOption.EC2APIEndpoint)
 
-	flags.Bool(operatorOption.AWSPaginationEnabled, true, "Enable pagination for AWS EC2 API requests. The default page size is 1000 items.")
+	flags.Int32(operatorOption.AWSMaxResultsPerCall, 0, "Maximum results per AWS API call for DescribeNetworkInterfaces and DescribeSecurityGroups. Set to 0 to let AWS determine optimal page size (default). If AWS returns OperationNotPermitted errors, automatically switches to 1000 for all future requests")
```
```suggestion
	flags.Int32(operatorOption.AWSMaxResultsPerCall, 0, "Maximum results per AWS API call for DescribeNetworkInterfaces and DescribeSecurityGroups. Set to 0 to let AWS determine optimal page size (default). If set to 0 and AWS returns OperationNotPermitted errors, automatically switches to 1000 for all future requests")
```

pillai-ashwin replied:
Done!

gandro commented on pkg/aws/ec2/ec2.go:
```diff
@@ -98,9 +97,53 @@ func NewClient(logger *slog.Logger, ec2Client *ec2.Client, metrics MetricsAPI, r
+
+	// Switch to pagination
+	c.maxResultsPerCall = 1000
+	c.logger.Warn("AWS API returned OperationNotPermitted, switching to paginated requests permanently",
```
Please add a hint in the log output that users shall set `aws-max-results-per-call` to 1000 to disable this warning.

pillai-ashwin replied:
Done!

gandro reviewed (commented):
Thanks! You missed one conversion with the retry loop, other than that no new feedback.

pillai-ashwin commented on pkg/aws/ec2/ec2.go:
```diff
@@ -942,25 +1014,37 @@ func createAWSTagSlice(tags map[string]string) []ec2_types.Tag {
+			if err != nil {
+				// If OperationNotPermitted and we switch to pagination, retry the entire request
+				if c.switchToPagination(err) {
+					break
```
Done!

gandro commented:
/test

gandro commented:
/test

pillai-ashwin commented:
/test

pillai-ashwin commented:
/test

pillai-ashwin commented:
/test

gandro commented:
@pillai-ashwin I think you have to rebase the branch again - the failures in E2E (complaining about an unknown flag) seem to indicate a mismatch between the test code (taken from main) and the code being tested in your branch.

liyihuang commented:
sorry, I thought I'm not reuiqred to approve it. I just removed myself

gandro commented:
/ci-l3-l4

gandro commented:
@pillai-ashwin You might have to rebase again to have the new ci-l7 ci-l3-l4 tests become green

pillai-ashwin commented:
/test
